fortran/plumed_f08.f90

```diff
@@ -2015,7 +2015,9 @@ end subroutine pl_cmd_ref_integer_0_0
     subroutine pl_cmd_ptr_integer_0_0(this,key,val,dummy,error)
       class(plumed),                 intent(inout) :: this ! inout to allow for initialization
       character(kind=c_char,len=*),  intent(in)    :: key
-      integer(KIND=c_int), pointer,  intent(in) :: val
+      ! workaround for bug in Intel 19 compiler (pointer => target)
+      ! https://github.com/dftbplus/dftbplus/commit/a9329e72791a224b005df70e7fe59dfb26b92a75
+      integer(KIND=c_int), target,  intent(in) :: val
```

Comment:
A few remarks on this:
* Indeed, unfortunately ifort 19 has problems with the original (though correct) formulation
* If you change `pointer` to `target`, you also must change `intent(in)` into `intent(inout)` as you change the value of the object (This was the typo, I mentioned earlier, which I had in my plumed interface in DFTB+)
* Although, the workaround is probably the closest you can get to the original formulation, they are not equivalent! The workaround does not force the caller to declare the `target`  attribute in the caller scope for the actual argument for `val` any more. Buy forgetting to do so, there is no warranty, that the compiler does not make a copy-in/copy-out when entering the `pl_cmd_ptr` routine. So, in theory, it could happen, that plumed obtains and stores a pointer to the address of a temporary object, which does not exist any more, when plumed tries to read or write it later. I am not aware of any compilers, which would do so (and it would be quite illogical), still it would be standard conforming to do so. :disappointed: 

I think, we need to add a few warnings/recommendation to the Fortran-interface somewhere in the docs. One for the `asynchronous` and one for the `target` attribute for the variables in the caller scope, when `pl_cmd_ptr` or `pl_cmd_const_ptr` is invoked.

Reply:
@aradi is there a standard preprocessor macro that can be used to have specific code for intel compiler? Was this bug submitted to intel already? Notice that on GH Actions we are using the latest intel compiler, so this is not just intel 19.

Reply:
No, the preprocessor is agnostic and does not know anything about the compiler which processes its output. The build framework should pass an appropriate define (e.g. `-DINTEL_FIX`) to the preprocessor, if it uses an Intel compiler, and then within the code one could do in `#:if`:
```fortran
#:if defined('INTEL_FIX')
  ...  ! target version
#:else
  ... ! pointer version
#:endif
```

As for reporting to intel, we did not report so far. Actually, I've just tried now to set up a minimal working example, and that actually works with Intel! I'll try a bit more, whether I can distill the problem. In case, you want to play around with it, my toy example was the following:
```
module test_target
  implicit none

  interface subptr
    module procedure subptr_integer, subptr_real
  end interface subptr

contains

  subroutine main()

    integer, target :: intval
    real, target :: realval

    intval = 42
    realval = 4E2

    call subptr(intval)

  end subroutine main


  subroutine subptr_integer(val)
    integer, pointer, intent(in) :: val

    print *, "INTEGER PTR obtained", val

  end subroutine subptr_integer


  subroutine subptr_real(val)
    real, pointer, intent(in) :: val

    print *, "REAL PTR obtained", val

  end subroutine subptr_real

end module test_target


program test_target_progam
  use test_target
  implicit none

  call main()

end program test_target_progam
```

Reply:
>If you change pointer to target, you also must change intent(in) into intent(inout) as you change the value of the object (This was the typo, I mentioned earlier, which I had in my plumed interface in DFTB+)

I didn't make this, and still the code is working. Is this expected?

You can see a sample code in the [regtest](https://github.com/plumed/plumed2/blob/576aca10897c539cbf7b4fc3a561c1770a65bf0a/regtest/basic/rt-make-fortran08/testfull.f90)

Reply:
And: shall I use `intent(in)` for `cmd_const_ptr` and `intent(inout)` for `cmd_ptr`?

Reply:
OK, I managed to create a minimal working example, and sent a [bug report](https://community.intel.com/t5/Intel-Fortran-Compiler/Compiler-bug-target-array-can-not-be-passed-for-pointer-argument/m-p/1327248#M158240) to the Intel forum.

Reply:
Yes, the code works irrespective of the `intent`. Since you pass the address of the object to something else, it does not really matter. Actually, now that I think once more about it, since you do not change the value of the object within the call itself, probably `intent(in)` is more appropriate, so leave it that way. What is more important, that the caller sets the `asynchronous` attribute in its scope in the non-constant case.

Reply:
> you do not change the value of the object within the call itself
This is not guaranteed in the `cmd_ptr` case. So I will use `intent(inout)` for that.

I mean: one should be able to use `cmd_ptr` in any place where it is possible to use any other `cmd_*` call, since it is the least restrictive one on the PLUMED side. So we cannot assume that PLUMED will not change the target.

Reply:
I see, that makes sense, indeed.

However, if possible, I would recommend to force the user to choose the most restrictive calling form. Calling `cmd_ptr` and `cmd_const_ptr` should **only** be done (and possible), if plumed stores a pointer and postpones the reading or writing. The caller is supposed to make a lot of extra gimics int those cases (`asynchronous`, `target`, etc.), which we can neither enforce nor check (either due to compiler bug or due to the lacking support in the Fortran standard). Therefore, the resulting code could in theory fail, and finding the bug would be a nightmare (try to find out that the segfault was due to the access to a non-existing temporary array...) In all other cases, when the action on the argument from Plumed happens only within the call, one should not allow the `cmd*ptr` calls, but force the user (via your run-time checking mechanism) to use `cmd_val` or `cmd_ref` instead. (By checking somewhere, whether the `nocopy` and eventually the `const` attributes were passed, and stop if they were not -- if that is possible on the C++ side. Then one given command had exactly one possible call).

Reply:
This would not be possible, and would not be compatible with the spirit of the original type-check implementation. The idea is to reproduce the behavior of type conversions in C++. If a given cmd requires a `const` reference, which means it promises that will not modify the buffer, a user should be allowed to pass a non `const` data.

> Therefore, the resulting code could in theory fail, and finding the bug would be a nightmare

This will not happen, unless users use `cmd_ptr` without `target`, which is not detectable by the compiler but explained in the documentation of the new interface. And, in any case, forcing the user to use `cmd_val` when PLUMED does not need the address will not help in this case.